src/main.rs

```diff
@@ -1,7 +1,160 @@
-use crossterm::event::{Event, KeyCode, KeyEvent, KeyModifiers};
-use crossterm::{event, terminal};
-use std::time::Duration;
 use anyhow::Result;
+use crossterm::cursor::{Hide, MoveTo, Show};
+use crossterm::event::{self, Event, KeyEvent};
+use crossterm::style::Print;
+use crossterm::terminal::{self, Clear, ClearType::*};
+use crossterm::queue;
+use std::io::{self, BufWriter, Stdout, Write};
+
+pub struct Kilo {
+    cursor: Cursor,
+    dimentions: Dimentions,
+    is_running: bool,
+    stdout: BufWriter<Stdout>,
+}
+
+impl Kilo {
+    pub fn new() -> Result<Self> {
+        Ok(Self {
+            cursor: Cursor::new(0, 0),
+            dimentions: Dimentions::from(terminal::size()?),
+            is_running: true,
+            stdout: BufWriter::new(io::stdout()),
+        })
+    }
+
+    pub fn run(&mut self) -> Result<()> {
+        let _override = RawModeOverride::new()?;
+
+        while self.is_running {
+            self.render()?;
+            self.process_input()?;
+        }
+
+        Ok(())
+    }
+
+    fn terminate(&mut self) -> Result<()> {
+        self.is_running = false;
+        queue!(self.stdout, Clear(All), MoveTo(0, 0))?;
```

Comment:
Should this go in a destructor instead?

Reply:
I'll try to rethink this. If anything, the problem is that this kind of does 2 things. The first one is it stops the main loop. And the second one is that it clears the screen. The second one might or might not go into the destructor. I'm not sure.